### PR TITLE
[FEATURE] alignment improvement refactorisation - free end gap initialisation

### PR DESCRIPTION
**Pull request by Irallia**
Resolves https://github.com/seqan/product_backlog/issues/71

### CODE REVIEW

File: include/seqan3/alignment/pairwise/detail/policy_optimum_tracker.hpp [hunk 1]
```diff
@@ -136,8 +174,10 @@ class policy_optimum_tracker
      * algorithm requires it, for example when a semi-global alignment shall be computed.
      */
     template <typename cell_t>
-    decltype(auto) track_last_column_cell(cell_t && cell, matrix_coordinate SEQAN3_DOXYGEN_ONLY(coordinate)) noexcept
+    decltype(auto) track_last_column_cell(cell_t && cell, matrix_coordinate coordinate) noexcept
     {
+        if (is_last_column_free)
```
Comment (joergi-w) [inline include/seqan3/alignment/pairwise/detail/policy_optimum_tracker.hpp:179]:
```suggestion
        if (last_column_is_free)
```

File: include/seqan3/alignment/pairwise/detail/policy_affine_gap_recursion.hpp [hunk 2]
```diff
@@ -149,7 +162,9 @@ class policy_affine_gap_recursion
      */
     affine_cell_type initialise_origin_cell() const noexcept
     {
-        return {score_type{}, gap_open_score, gap_open_score};
+        return {0,
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/detail/policy_affine_gap_recursion.hpp:165]:
```suggestion 
        return {score_type{},
```

File: include/seqan3/alignment/pairwise/detail/policy_optimum_tracker.hpp [hunk 3]
```diff
@@ -57,6 +58,34 @@ class policy_optimum_tracker
...
+     *
+     * \param[in] cell The current cell to be tracked.
+     * \param[in] coordinate The matrix coordinate of the current cell.
+     *
+     * \returns The forwarded cell.
+     *
+     * \details
+     *
+     * A call to this function only tracks the optimal score of the given cell if the configuration of the alignment
+     * algorithm requires it. This is the case whenever we want to get the path through the matrix, for example to print
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/detail/policy_optimum_tracker.hpp:78]: I would remove the last sentence.

File: include/seqan3/alignment/pairwise/detail/policy_optimum_tracker.hpp [hunk 4]
```diff
@@ -116,8 +151,10 @@ class policy_optimum_tracker
      * algorithm requires it, for example when a semi-global alignment shall be computed.
      */
     template <typename cell_t>
-    decltype(auto) track_last_row_cell(cell_t && cell, matrix_coordinate SEQAN3_DOXYGEN_ONLY(coordinate)) noexcept
+    decltype(auto) track_last_row_cell(cell_t && cell, matrix_coordinate coordinate) noexcept
     {
+        if (last_row_is_free)
+            find_new_optimum(cell, std::move(coordinate));
         return std::forward<cell_t>(cell);
     }
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/detail/policy_optimum_tracker.hpp:159]:
```suggestion
            find_new_optimum(cell, std::move(coordinate));

        return std::forward<cell_t>(cell);
    }
```

File: include/seqan3/alignment/pairwise/detail/policy_optimum_tracker.hpp [hunk 5]
```diff
@@ -136,8 +173,10 @@ class policy_optimum_tracker
      * algorithm requires it, for example when a semi-global alignment shall be computed.
      */
     template <typename cell_t>
-    decltype(auto) track_last_column_cell(cell_t && cell, matrix_coordinate SEQAN3_DOXYGEN_ONLY(coordinate)) noexcept
+    decltype(auto) track_last_column_cell(cell_t && cell, matrix_coordinate coordinate) noexcept
     {
+        if (last_column_is_free)
+            find_new_optimum(cell, std::move(coordinate));
         return std::forward<cell_t>(cell);
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/detail/policy_optimum_tracker.hpp:180]:
```suggestion
            find_new_optimum(cell, std::move(coordinate));

        return std::forward<cell_t>(cell);
```

File: include/seqan3/alignment/pairwise/detail/policy_optimum_tracker.hpp [hunk 6]
```diff
@@ -158,10 +197,8 @@ class policy_optimum_tracker
     template <typename cell_t>
     decltype(auto) track_final_cell(cell_t && cell, matrix_coordinate coordinate) noexcept
     {
-        bool const is_better_score = cell.optimal_score() >= optimal_score;
-        optimal_score = (is_better_score) ? cell.optimal_score() : optimal_score;
-        optimal_coordinate = (is_better_score) ? std::move(coordinate) : optimal_coordinate;
-
+        if (!(last_column_is_free || last_row_is_free))
+            find_new_optimum(cell, std::move(coordinate));
         return std::forward<cell_t>(cell);
```
Comment (rrahn) [inline include/seqan3/alignment/pairwise/detail/policy_optimum_tracker.hpp:202]:
```suggestion
            find_new_optimum(cell, std::move(coordinate));

        return std::forward<cell_t>(cell);
```

**Pull request merged**